cli/paas/install_client.go

```diff
@@ -21,88 +29,99 @@ type InstallClient struct {
-	// function which queries the cluster for the necessary
-	// data. If that data could not be found the system will fall
-	// back to cli option and/or interactive entry by the user.
-	//
-	// NOTE: This is function is set here and not in the gitea
-	// definition because the function has to have access to the
-	// cluster in question, and that information is only available
-	// now, not at deployment declaration time.
-
-	domain, err := deployments.NeededOptions.GetOpt("system_domain", "")
+	var err error
+	options, err = options.Populate(kubernetes.NewCLIOptionsReader(cmd))
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
-	domain.DynDefaultFunc = func(o *kubernetes.InstallationOption) error {
-		ips := c.kubeClient.GetPlatform().ExternalIPs()
-		if len(ips) > 0 {
-			domain := fmt.Sprintf("%s.nip.io", ips[0])
-			o.Value = domain
-		}
-		// else leave invalid, to be handled by cli option
-		// reader or interactive entry
-		return nil
-	}
-
-	deployments.PopulateNeededOptions(kubernetes.NewCLIOptionsReader(cmd))
 
 	interactive, err := cmd.Flags().GetBool("interactive")
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
 
 	if interactive {
-		deployments.PopulateNeededOptions(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		options, err = options.Populate(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	} else {
-		deployments.PopulateNeededOptions(kubernetes.NewDefaultOptionsReader())
+		options, err = options.Populate(kubernetes.NewDefaultOptionsReader())
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	}
 
-	c.showInstallConfiguration(deployments)
+	c.showInstallConfiguration(options)
 
 	// TODO (post MVP): Run a validation phase which perform
 	// additional checks on the values. For example range limits,
 	// proper syntax of the string, etc. do it as pghase, and late
 	// to report all problems at once, instead of early and
 	// piecemal.
 
-	err = c.deploySet(deployments)
+	deployment := deployments.Traefik{Timeout: DefaultTimeoutSec}
+	deployment.Deploy(c.kubeClient, c.ui, options.ForDeployment(deployment.ID()))
+	if err != nil {
+		return err
+	}
+
+	// Try to give a nip.io domain if the user didn't specify one
+	domain, err := options.GetOpt("system_domain", "")
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
 
-	c.ui.Success().Msg("Carrier installed.")
+	err = c.fillInMissingSystemDomain(domain)
+	if err != nil {
+		return errors.Wrap(err, "Couldn't install carrier")
+	}
+	if domain.Value.(string) == "" {
+		return errors.New("You didn't provide a system_domain and we were unable to setup a nip.io domain (couldn't find and ExternalIP)")
+	}
+
+	c.ui.Success().Msg("Created system_domain: " + domain.Value.(string))
+
+	for _, deployment := range []kubernetes.Deployment{
```

Comment:
The only thing I would do now, would be moving these slices, and the single deployments to the top of the file as variables.
That is most definitely a nit we can defer for a while.